Automattic/newspack-newsletters
# feat: refactor add_contact and put async sign up behind a flag

### All Submissions:






### Changes proposed in this Pull Request:



Taking baby steps to centralize our data flows by deprecating some methods and moving them to a new class.

At this point there's no relevant changes made, it's mostly moving code around. In a next PR we are going to start changing things, and also making changes to the places where this deprecated methods are called.

The only change we make is to put the Async strategy behind a constant.

### How to test the changes in this Pull Request:

1. Test signing up for a newsletter using the Newsletters Subscription block
2. Confirm it works as expected (check the ESP dashboard to make sure the contact was added to the correct lists)
3. Confirm you see `PHP Deprecated:  Function Newspack_Newsletters_Subscription::add_contact is <strong>deprecated</strong> since version 2.21! Use Newspack_Newsletters_Contacts::upsert instead. in /var/www/html/wp-includes/functions.php on line 6085` on your logs
4. Go to my account, verify your email, and test updating your selection of newsletters lists. Confirm it works
5. Confirm you see `PHP Deprecated:  Function Newspack_Newsletters_Subscription::add_contact is <strong>deprecated</strong> since version 2.21! Use Newspack_Newsletters_Contacts::upsert instead. in /var/www/html/wp-includes/functions.php on line 6085`
6. As an admin, delete the user
7. Confirm the user was deleted from the ESP
8. Add `define( 'NEWSPACK_NEWSLETTERS_ASYNC_SUBSCRIPTION_ENABLED', true );` to `wp-config.php` and repeat the tests, making sure the async strategy is still working


### Other information:





---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207717004877598
